Automattic/woocommerce-payments
# Introduce express checkout utilities class

Fixes #7694
Fixes #7801

#### Changes proposed in this Pull Request

This introduces a utility class for express checkout buttons such as the payment request buttons and WooPay buttons. There are several duplicated functions across the WooPay and Payment Request button handler class. The impetus for this refactor was the bug mentioned in #7588. Both express buttons are using the `wc_ajax_wcpay_add_to_cart` hook with an attached `ajax_add_to_cart` function. This leads to conflicts when both button types are enabled at the same time, with the WooPay `wc_ajax_wcpay_add_to_cart` always being called. This PR moves that hook to a single location and calls a single `wc_ajax_wcpay_add_to_cart` function from a utilities class.

This will also serve as a good base to consolidate other duplicated functions among the two button handler classes as mentioned in 1545-gh-Automattic/woopay. 









#### Testing instructions





1. While on the `develop` branch, enable WooPay and Payment Request buttons.
2. Navigate to a shippable product and ensure both buttons appear on the page.
3. Increase the quantity of the product to something more than 1.
4. Click the GPay button and notice that the total does not reflect the quantity entered and reflects only a quantity of 1
5. Apply this PR and re-test. The total should reflect the selected product quantity.
6. Open the network tab and check that no errors are occurring in the`wc-ajax=wcpay_add_to_cart` and `wc-ajax=wcpay_get_shipping_options` responses.
7. Check that orders are completing as expected for both buttons.

I remember having trouble reproducing this initially due my my shipping setup. As long as you're triggering a shipping zone/method with an address stored in Google, the issue should be present. I'm using a flat rate of $1. 

-------------------


**Post merge**
